Fix SQLite panic when syncing many UTXOs

notmandatory opened the pull request:
### Description

fixes #1827 

I changed the `SqliteDatabase` struct to contain a `r2d2` with [`r2d2_sqlite`](https://docs.rs/r2d2_sqlite/latest/r2d2_sqlite/) connection pool instead of a sqlite `Connection` so the `SqliteDatabase` can be safely cloned and reused when a batch is started. This prevents a large sync from trying to start and commit a batch (db transaction) while the initial non-batch connection is still busy writing it's data.

### Notes to the reviewers

I had to comment out the `test_batch_script_pubkey()` test for `sqlite` since the fix in this PR prevents a new sqlite DB connections from being created when a batch is started. In practice our sync'ing code does all its initial work without starting a batch (db transaction), then before finishing it starts a batch to do some cleanup work and commits it.

### Changelog notice

- Fix SQLite panic when syncing many UTXOs #1827

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### Bugfixes:

* [ ] This pull request breaks the existing API
* [x] I've added tests to reproduce the issue which are now passing
* [x] I'm linking the issue being fixed by this PR

notmandatory commented:
Test completes without panic in ~58 mins.
```
test blockchain::electrum::test::test_electrum_large_num_utxos ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 327 filtered out; finished in 3501.34s
```

notmandatory commented on src/database/sqlite.rs:
```diff
@@ -76,10 +76,8 @@ static MIGRATIONS: &[&str] = &[
 #[derive(Debug)]
 pub struct SqliteDatabase {
-    /// Path on the local filesystem to store the sqlite file
-    pub path: PathBuf,
```
I removed this since it's no longer needed.

notmandatory commented:
Rebased on release/0.30 branch.

notmandatory committed:
ci: fix pinned rustls and add ci/pin-msrv.sh

notmandatory commented:
The sync is still very slow, but that's due to how  electrum syncing works.  It requests complete histories for all  addresses for every sync. To support these sort of very large UTXO wallets on a mobile device I think a protocol like CBF (kyoto) is the only way to do it. CBF syncs block by block, so once blocks with new TX are processed they don't get processed again. Down side of CBF is it doesn't check the mempool, but a way to combine CBF for committed TX and electrum or esplora to get unconfirmed mempool TX should work.

ankitgoswami commented on src/database/sqlite.rs:
```diff
@@ -76,10 +74,8 @@ static MIGRATIONS: &[&str] = &[
-    pub path: PathBuf,
     /// A rusqlite connection object to the sqlite database
-    pub connection: Connection,
+    pub connection: Rc<Connection>,
```
Could Arc be used instead to maintain thread safety?

notmandatory replied:
I did a test with `Arc<RwLock<Connection>>` but this causes deadlocks. The `rusqlite::Connection` type is a little strange in that it is `Send` but that guarantee is provided by the underlying C library not by Rust. I ended up using the `r2d2_sqlite` connection pool which supports thread safety and cloning, which is what we needed.

nymius commented:
I'm trying to run `test_electrum_large_num_utxos`, reducing `NUM_TX` to 10, as you mentioned in [this comment](https://github.com/bitcoindevkit/bdk/issues/1827#issuecomment-2646023260), but I'm getting dependency errors. Could you provide more insight about how to test it? I think I missing an important part of the setup.

reez commented:
Just adding some info when pointing to this branch from bdk-ffi and building.

I'm using a branch based on the `bdk-ffi` 0.31.2 tag.

That branch was pointing to `bdk` version 0.29.0, so I pointed to this PR branch instead `"https://github.com/notmandatory/bdk", branch = "fix/many_utxos"`

I won't list every error because from what I can tell the errors all seem to stem from the same root cause of Rc<Connection> not being Send + Sync, but definitely let me know if you'd like to me to follow up with them.

```
bdk-ffi % cargo build

error[E0277]: `Rc<Connection>` cannot be sent between threads safely
     |
1461 | uniffi::deps::static_assertions::assert_impl_all!(r#Wallet: Sync, Send);
     |                                                   ^^^^^^^^ `Rc<Connection>` cannot be sent between threads safely
     |
     = help: within `AnyDatabase`, the trait `std::marker::Send` is not implemented for `Rc<Connection>`, which is required by `wallet::Wallet: Sync`
```

notmandatory committed:
fix(sqlite): use clonable r2d2 sqlite connection pool in SqliteDatabase

    This prevents: Error { code: DatabaseBusy, extended_code: 5 }, Some("database is locked")
    which occured when syncing very large number of utxos.
    See: electrum::test::test_electrum_large_num_utxos

notmandatory commented:
I'm replacing the `Rc` with a `r2d2::Pool` which is Send + Sync and builds with `bdk-ffi`.  But I'm also having trouble re-running the test so trying to figure that out now.

notmandatory committed:
ci: downgrade electrsd dev dep to 0.24.0

notmandatory committed:
release: bump version to 0.30.2

notmandatory commented:
I retested with the `r2d2` connection pool and test passes with no errors: 
```
test blockchain::electrum::test::test_electrum_large_num_utxos ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 327 filtered out; finished in 3427.42s
```

I also verified `bdk-ffi` builds with no errors. We just need to add one new `bdk-ffi::BdkError` enum variant in the UDL:
```
[Error]
enum BdkError {
  ...
  "R2d2",
};
```

notmandatory commented:
> I'm trying to run `test_electrum_large_num_utxos`, reducing `NUM_TX` to 10, as you mentioned in [this comment](https://github.com/bitcoindevkit/bdk/issues/1827#issuecomment-2646023260), but I'm getting dependency errors. Could you provide more insight about how to test it? I think I missing an important part of the setup.

You should be able to test it now. I think I forgot to check in some dependency fixes that were preventing the electrum tests from building.

nymius commented:
> You should be able to test it now. I think I forgot to check in some dependency fixes that were preventing the electrum tests from building.

Thanks! It works now

nymius commented on src/blockchain/electrum.rs:
```diff
@@ -429,4 +427,179 @@ mod test {
+
+    #[cfg(feature = "sqlite")]
+    #[test]
+    #[ignore] // takes ~1 hr to complete, here as reference for future testing
```
Should we remove this comment or update it? With `NUM_TX = 2` (as it is now) it took ~106 seconds on average to run for my computer (I tested it multiple times).

notmandatory replied:
oh ya I forgot to change it back from 2, it might take 10 to show the error or more on a faster computer. I'm going to see if I can repro the error with the test but not the fix from this PR but with the fix in #1831 instead.

notmandatory replied:
I also had a test run that didn't fail with 5 but did with 10 tx. I put the test back to 50 anyway.

nymius commented on src/blockchain/electrum.rs:
```diff
@@ -429,4 +427,179 @@ mod test {
+            .unwrap();
+        assert_eq!(wallet1.get_balance().unwrap().confirmed, 5_0000_0000);
+        // bdk wallet 1 creates NUM_TX tx * NUM_UTXO utxos and sends them back to itself
+        for _ in 0..NUM_TX {
```
If not so time consuming I would implement these as inner functions `fund_wallet_with_many_utxos` and `send_many_utxos`. Will improve discoverability of the code for future occasions if we have to reproduce a similar scenario.

notmandatory replied:
I'll work on a cleaner version for the 1.0 master branch since this 0.30.x branch of code will be ending soon.

nymius reviewed (commented):
tACK

Didn't manage to produce the failure at commit a75f749 using `test_electrum_large_num_utxos` with `NUM_TX = 5`

reez commented:
> I also verified `bdk-ffi` builds with no errors. We just need to add one new `bdk-ffi::BdkError` enum variant in the UDL:

I verified `bdk-ffi` builds w no errors as well, and also built the `bdk-swift` bindings locally & used those local bindings in BDK Swift Example Wallet.

jguinta committed:
Address out of memory issue caused by many txs.

For some wallets there exists a pathological case where we may try to fetch many thousands
        of transactions at once, which creates enormous memory pressure. By chunking the batch
        into more reasonably sized sub-queries, we allow time for memory to be freed.

notmandatory commented:
I rebased this PR on #1831 which fixes another issue related to syncing large transactions. After #1831 is merged this one can go in.

thunderbiscuit commented:
We'll be ready to build a patch release of the bdk-ffi libraries once both those fixes are in.

evanlinjin commented on src/database/sqlite.rs:
```diff
@@ -919,23 +925,20 @@ impl BatchDatabase for SqliteDatabase {
-        let db = SqliteDatabase::new(self.path.clone());
-        db.connection.execute("BEGIN TRANSACTION", []).unwrap();
-        db
+        let connection = self.pool.get().unwrap();
+        connection.execute("BEGIN TRANSACTION", []).unwrap();
+        SqliteDatabase {
+            pool: self.pool.clone(),
+        }
     }
 
     fn commit_batch(&mut self, batch: Self::Batch) -> Result<(), Error> {
-        batch.connection.execute("COMMIT TRANSACTION", [])?;
+        let connection = batch.pool.get()?;
+        connection.execute("COMMIT TRANSACTION", [])?;
```
From my understanding, an [sqlite transaction is isolated to a single connection](https://sqlite.org/isolation.html), unless we are in "shared cache mode".

~Even in "shared cache mode", is it even possible to start a transaction in one connection and commit it in another?~ Edit: It is definitely not possible. [Almost all statements in sqlite are automatically a transaction](https://sqlite.org/lang_transaction.html) so it would make no sense that transactions are begun and committed from different connections.

notmandatory replied:
Yes by default all the individual statements in the syncing happen in their on transaction. But by starting and ending a db transaction it ensures that a group of statements are committed or rolled back atomically. For this final phase of syncing `into_db_update` and also in `cache_addresses` there are multi statements that need to commit or roll back together. 

The main thing I'm fixing by introducing this `r2d2` pool is that we don't have different threads making their own `Connection` to the same sqlite DB file. Even when the `SqliteDatabase.pool` is cloned it's still using the same `Connection`. There's probably a way to do this without using a connection pool like `r2d2` but the lifetimes and ownership gets hairy and I didn't want to roll my own when a common solution is to use a pool.

notmandatory replied:
~~For reference, the `r2d2_sqlite` connection manager uses a `Mutex<Option<Connection>>` so that a single connection can be shared. I could remove `r2d2` and do the same but then I'd have to replicate the connection logic they already have.~~ Sorry this is wrong, the Mutex is only used for sqlite in memory connections.

https://github.com/ivanceras/r2d2-sqlite/blob/fbba255ff23aeaa005e810ce16fae19f1e44b0c2/src/lib.rs#L60-L65

evanlinjin replied:
@notmandatory Okay that makes sense, thanks for clarifying! Maybe a comment somewhere stating that `ConnectionPool` only manages a single connection will be helpful.

evanlinjin replied:
I guess we can create an `r2d2` Pool with the `max_size` set to 1.

notmandatory replied:
I'm trying to find out exactly what `r2d2` does but it defaults to 10 connections and everything still works. I think the reason is `sqlite` allows multiple connections on the same file but only a single "writer" as described by https://github.com/rusqlite/rusqlite/issues/920. I think `r2d2` is doing something else to avoid the "DatabaseBusy" error. The `r2d2_sqlite` crate has a `is_valid()` function that starts an empty "batch" which I believe fails if the connection is not writable. https://github.com/ivanceras/r2d2-sqlite/blob/fbba255ff23aeaa005e810ce16fae19f1e44b0c2/src/lib.rs#L163C5-L165C6

notmandatory replied:
Now I believe we actually don't want a pool of `max_size` 1 since it will block safe read-only queries we make during write transactions.

evanlinjin reviewed (approved):
ACK 29f6935b440762af6768eaddd127435b4b088587

Only request: Comment somewhere saying that the Sqlite Connection Pool will always only have a single connection.

evanlinjin reviewed (changes requested):
Need to ensure we handle db transactions properly (as discussed above).

evanlinjin commented on src/database/sqlite.rs:
```diff
@@ -919,23 +925,20 @@ impl BatchDatabase for SqliteDatabase {
     type Batch = SqliteDatabase;
 
     fn begin_batch(&self) -> Self::Batch {
-        let db = SqliteDatabase::new(self.path.clone());
```
I think a more straight-forward fix would be to just somehow clone `self` and return it. Trying it out now.

notmandatory replied:
Unfortunately `rusqlite::Connection` doesn't support `Clone`. I tried `Arc<Mutex>` and separating read and write connections, including setting the sqlite connection flag on the read connection to be READ_ONLY, but continue to get the original err.

I'd like to stick with this pool solution since I don't think there's any reason to doubt `r2d2` handles multiple concurrent transactions, even if I can't pinpoint exactly how. Moreover by using this connection pool 1) our tests pass, 2) the `bdk_ffi` project builds and bitkey's app doesn't crash on their example big wallet 3) the `r2d2_sqlite` example demos multiple concurrent threads writing to the DB in their own automatic transactions. Bottom line is I don't want to re-invent a multi-threaded sqlite connection pool when we already have one that works.

As a side note, I did test with the `r2d2` pool configured to use only one thread and my test still passes, but I'd rather leave it at the default max 10 connections so we don't unnecessarily limit performance.

notmandatory commented:
Replaced by #1836
